server/plugin/api.go

```diff
@@ -110,101 +116,119 @@ func (p *MatterpollPlugin) handleLogo(w http.ResponseWriter, r *http.Request) {
 	http.ServeFile(w, r, filepath.Join(bundlePath, "assets", iconFilename))
 }
 
-func (p *MatterpollPlugin) handleVote(w http.ResponseWriter, r *http.Request) {
-	vars := mux.Vars(r)
-	pollID := vars["id"]
-	optionNumber, _ := strconv.Atoi(vars["optionNumber"])
-	response := &model.PostActionIntegrationResponse{}
+func (p *MatterpollPlugin) handlePostActionIntegrationRequest(handler postActionHandler) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		request := model.PostActionIntegrationRequestFromJson(r.Body)
+		if request == nil {
+			p.API.LogWarn("failed to decode PostActionIntegrationRequest")
+			http.Error(w, "invalid request", http.StatusBadRequest)
+			return
+		}
+		userLocalizer := p.getUserLocalizer(request.UserId)
 
-	request := model.PostActionIntegrationRequestFromJson(r.Body)
-	if request == nil {
-		w.WriteHeader(http.StatusBadRequest)
-		return
+		msg, update, err := handler(mux.Vars(r), request)
+		if err != nil {
+			p.API.LogWarn("failed to handle PostActionIntegrationRequest", "error", err.Error())
+		}
+
+		response := &model.PostActionIntegrationResponse{}
+		if msg != nil {
+			response.EphemeralText = p.LocalizeDefaultMessage(userLocalizer, msg)
+		}
+		if update != nil {
+			response.Update = update
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		if _, err = w.Write(response.ToJson()); err != nil {
+			p.API.LogWarn("failed to write PostActionIntegrationResponse", "error", err.Error())
+		}
+	}
+}
+
+func (p *MatterpollPlugin) handleSubmitDialogRequest(handler submitDialogHandler) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		request := model.SubmitDialogRequestFromJson(r.Body)
+		if request == nil {
+			p.API.LogWarn("failed to decode SubmitDialogRequest")
+			http.Error(w, "invalid request", http.StatusBadRequest)
+			return
+		}
+
+		msg, response, err := handler(mux.Vars(r), request)
+		if err != nil {
+			p.API.LogWarn("failed to handle SubmitDialogRequest", "error", err.Error())
+		}
+
+		if msg != nil {
+			userLocalizer := p.getUserLocalizer(request.UserId)
+			p.SendEphemeralPost(request.ChannelId, request.UserId, p.LocalizeDefaultMessage(userLocalizer, msg))
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		if _, err = w.Write(response.ToJson()); err != nil {
+			p.API.LogWarn("failed to write PostActionIntegrationResponse", "error", err.Error())
+		}
 	}
+}
+
+func (p *MatterpollPlugin) handleVote(vars map[string]string, request *model.PostActionIntegrationRequest) (*i18n.Message, *model.Post, error) {
```

Comment:
With this style, we can't use `Mattermost-User-Id` header.  So `postActionHandler` and `submitDialogHandler` should have `userId` or `http.Header` as arguments.

Reply:
I have thought about this and I have a solution that works with this code. Would you be fine if we discuss this in a separate PR?

Reply:
Sure 👍